src/Caching/Impl/ModuleDatabase.cs

```diff
@@ -0,0 +1,156 @@
+                            return ModuleStorageState.Corrupted;
+                        }
+
+                        var modules = db.GetCollection<ModuleModel>("modules");
+                        var model = modules.Find(m => m.Name == moduleName).FirstOrDefault();
+                        if (model == null) {
+                            return ModuleStorageState.DoesNotExist;
+                        }
+
+                        module = new PythonDbModule(model, filePath, _services);
+                        return ModuleStorageState.Complete;
+                    }
+                } catch (Exception ex) when (ex is IOException || ex is UnauthorizedAccessException) {
+                    Thread.Sleep(10);
+                }
+            }
+            return ModuleStorageState.DoesNotExist;
+        }
+
+        public Task StoreModuleAnalysisAsync(IDocumentAnalysis analysis, CancellationToken cancellationToken = default)
+            => Task.Run(() => StoreModuleAnalysis(analysis, cancellationToken));
+
+        private void StoreModuleAnalysis(IDocumentAnalysis analysis, CancellationToken cancellationToken = default) { 
+            var model = ModuleModel.FromAnalysis(analysis, _services);
+            Exception ex = null;
+            for (var retries = 50; retries > 0; --retries) {
+                cancellationToken.ThrowIfCancellationRequested();
+                try {
+                    if (!_fs.DirectoryExists(_databaseFolder)) {
+                        _fs.CreateDirectory(_databaseFolder);
+                    }
+
+                    cancellationToken.ThrowIfCancellationRequested();
+                    using (var db = new LiteDatabase(Path.Combine(_databaseFolder, $"{model.UniqueId}.db"))) {
+                        var modules = db.GetCollection<ModuleModel>("modules");
+                        modules.Upsert(model);
+                        return;
+                    }
+                } catch (Exception ex1) when (ex1 is IOException || ex1 is UnauthorizedAccessException) {
+                    ex = ex1;
+                    Thread.Sleep(10);
+                } catch (Exception ex2) {
+                    ex = ex2;
+                    break;
+                }
+            }
+
+            if (ex != null) {
+                _log?.Log(System.Diagnostics.TraceEventType.Warning, $"Unable to write analysis of {model.Name} to database. Exception {ex.Message}");
+                if (ex.IsCriticalException()) {
+                    throw ex;
+                }
+            }
+        }
+
+        /// <summary>
+        /// Locates database file based on module information. Module is identified
+        /// by name, version, current Python interpreter version and/or hash of the
+        /// module content (typically file sizes).
+        /// </summary>
+        private string FindDatabaseFile(string moduleName, string filePath) {
+            var interpreter = _services.GetService<IPythonInterpreter>();
```

Comment:
So now we always have one interpreter per LS, right?

Reply:
We always actually had... There was notion to have multiple service containers for multi-root workspaces but how actually this is handled is that extension runs multiple language servers, one per root.

Reply:
Ok